// virtual-device-sdk/lib/src/VirtualDevice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const https = require("https");
class VirtualDevice {
    constructor(token) {
        this.token = token;
        this.baseURL = "https://virtual-device.bespoken.io/process";
    }
    normalizeMessage(message) {
        if (message.trim().toLowerCase() === "no") {
            return "alexa no";
        }
        return message;
    }
    message(message, debug) {
        message = this.normalizeMessage(message);
        let url = this.baseURL + "?message=" + message + "&user_id=" + this.token;
        if (debug) {
            url += "&debug=true";
        }
        const promise = new Promise((resolve, reject) => {
            const callback = (response) => {
                let data = "";
                response.on("data", (chunk) => {
                    data += chunk;
                });
                response.on("end", () => {
                    if (response.statusCode === 200) {
                        const result = JSON.parse(data);
                        result.transcript = this.normalizeTranscript(result.transcript);
                        result.message = message;
                        resolve(result);
                    }
                    else {
                        reject(data);
                    }
                });
            };
            const request = https.get(url, callback);
            request.on("error", function (error) {
                reject(error);
            });
            request.end();
        });
        return promise;
    }
    resetSession() {
        return this.message("alexa quit");
    }
    normalizeTranscript(transcript) {
        if (!transcript) {
            return null;
        }
        return transcript.toLowerCase();
    }
}
exports.VirtualDevice = VirtualDevice;
